File: implement-shell-tools/ls/my_ls.js
const fs = require('fs');
const path = require('path');

const args = process.argv.slice(2);

let showHidden = false;
let targetDir = '.';

// Parse arguments
for (let i = 0; i < args.length; i++) {
    if (args[i] === '-a') {
        showHidden = true;
    } else if (args[i] === '-1') {
        // -1 just means one per line, which we always do – nothing to store
    } else {
        // Assume it's the directory path
        targetDir = args[i];
    }
}

// Read directory contents
let entries = fs.readdirSync(targetDir);

// Filter hidden files unless -a is used
if (!showHidden) {
    entries = entries.filter(entry => !entry.startsWith('.'));
}

// Sort alphabetically
entries.sort();

// Print one per line
entries.forEach(entry => console.log(entry));

File: implement-shell-tools/ls/my_ls.py
import os
import sys

args = sys.argv[1:]

showHidden = False
targetDir = '.'

i = 0
while i < len(args):
    if args[i] == '-a':
        showHidden = True
    elif args[i] == '-1':
        # one per line is default so nothing to do
        pass
    else:
        targetDir = args[i]
    i = i + 1

entries = os.listdir(targetDir)

if not showHidden:
    newEntries = []
    for entry in entries:
        if not entry.startswith('.'):
            newEntries.append(entry)
    entries = newEntries

entries.sort()

for entry in entries:
    print(entry)